frontend/app/scripts/react-directives/device-manager/config-editor/deviceTabStore.js

```diff
@@ -159,40 +194,138 @@ export class EmbeddedSoftware {
+      component.resetUpdate();
     }
   }
 
   async startFirmwareUpdate(address, portConfig) {
-    await this.firmware.startUpdate(address, portConfig);
+    const componentsCanBeUpdated = this.componentsCanBeUpdated;
+
+    try {
+      this.firmware.setupUpdate();
+      this.setupComponentsUpdate(componentsCanBeUpdated);
+      await this.firmware.startUpdate(address, portConfig, componentsCanBeUpdated);
+    } catch (err) {
+      this.firmware.resetUpdate();
+      this.resetComponentsUpdate(componentsCanBeUpdated);
+      throw err;
+    }
+    this.clearComponentsVersion();
   }
 
   async startBootloaderUpdate(address, portConfig) {
-    await this.bootloader.startUpdate(address, portConfig);
+    const componentsCanBeUpdated = this.componentsCanBeUpdated;
+
+    try{
+      this.bootloader.setupUpdate();
+      this.firmware.setupUpdate();
+      this.setupComponentsUpdate(componentsCanBeUpdated);
+      await this.bootloader.startUpdate(address, portConfig, componentsCanBeUpdated);
+    } catch (err) {
+      this.bootloader.resetUpdate();
+      this.firmware.resetUpdate();
+      this.resetComponentsUpdate(componentsCanBeUpdated);
+      throw err;
+    }
     this.firmware.clearVersion();
+    this.clearComponentsVersion();
+  }
+
+  async startComponentsUpdate(address, portConfig) {
+    const componentsCanBeUpdated = this.componentsCanBeUpdated
+
+    try{
+      this.setupComponentsUpdate(componentsCanBeUpdated);
+      const firstComponent = componentsCanBeUpdated.get(0);
+      // Triggering update for the first component with components list causes all components update
+      await firstComponent.startUpdate(address, portConfig, componentsCanBeUpdated);
+    } catch (err) {
+      this.resetComponentsUpdate(componentsCanBeUpdated);
+      throw err;
+    }
   }
 
   clearVersion() {
     this.firmware.clearVersion();
     this.bootloader.clearVersion();
+    this.clearComponentsVersion();
   }
 
   clearError() {
     this.firmware.clearError();
     this.bootloader.clearError();
+    this.clearComponentsError();
   }
 
   get isUpdating() {
-    return this.firmware.isUpdating || this.bootloader.isUpdating;
+    if (this.firmware.isUpdating || this.bootloader.isUpdating) {
+      return true;
+    }
+    for (const component of this.components.values()) {
+      if (component.isUpdating) {
+        return true;
+      }
+    }
+    return false;
   }
 
   get hasUpdate() {
-    return this.firmware.hasUpdate;
+    if (this.firmware.hasUpdate) {
+      return true;
+    }
+    return this.hasComponentsUpdates;
+  }
+
+  get hasComponentsUpdates() {
+    return this.componentsCanBeUpdated.size > 0;
+  }
+
+  get componentsCanBeUpdated() {
+    return new Map([...this.components].filter(([, component]) => component.hasUpdate));
   }
 
   get hasError() {
-    return this.firmware.hasError || this.bootloader.hasError;
+    if (this.firmware.hasError || this.bootloader.hasError) {
```

Comment:
`return this.firmware.hasError || this.bootloader.hasError || [...this.components.values()].some(c => c.hasError);`